garb/garb_recv_loop.cpp

```diff
@@ -156,12 +161,51 @@ RecvLoop::one_loop()
+                                //    waiting in TCP connection
+                                // 2. SST finished, script received data
+                                // Because of the socat problem, if we end up in case 1, we have to
+                                // send kill to the whole group (script and its children) to avoid
+                                // socat hanging forever and blocking ports for next requests.
+                                // In case 1 we should simply wait for script to finish.
+                                // Unfortunately from garbd point of view it is not trivial to distinguish
+                                // these two cases. That's why we always terminate the whole group
+                                // just after receiving sst.
+                                // If any post-processing of received SST has to be done, it should be done
+                                // in post-sst-script.
+                                //
+                                // Having said that, we will end up in 'if (sst_terminated_)' or
+                                // 'else if(sst_ended_)' branch below randomly, (race between err_log_thd
+                                // and sst_status_thd which is not good in general, but it is as it is.
+
+                                // Becaus of the above, we will introduce a little ugly hack here.
+                                // After successful SST Donor exits from 'donor' state immediately.
+                                // That means we've got a race condition between err_log_thd and sst_status_thd.
+                                // Whatever happens first:
+                                // 1. script exits and we catch it in err_log_thd
+                                // 2. Donor exits from 'donor' state and we catch it in sst_status_thd
+                                // we will either wait for script to finish or terminate it with pgkill
+                                // There is no easy and reliable way to synchronize it, because from garbd point
+                                // of view we don't have the information about SST success/failure when handling
+                                // event of Donor moving from 'donor' state.
+                                // The hack is to wait 'a bit' to let sst-script to finish gracefully and then decide
+                                // if we still need to kill it or not.
+                                // It is up to SST script to exit immediately after receiving SST.
+                                // If any post processing has to be done it has to be done in post-sst-script.
+
+                                // Wait up to 5 seconds for recv-script to finish
+                                std::unique_lock<std::mutex> lock(script_end_mtx_);
+                                script_end_cv_.wait_for(lock, std::chrono::seconds(5));
```

Comment:
Technically feasible, but I don't want to do this. This will open the gate for someone who wants to do long post processing in sst script. Instead of doing this, post processing should be done in post-sst script. 5 secs is dirty hack to avoid 2-way execution synchronization implementation between sst-script and garbd.